src/vsr/replica.zig

```diff
@@ -2335,49 +2464,86 @@ pub fn Replica(
         fn commit_pipeline(self: *Self) void {
             assert(self.status == .normal);
             assert(self.leader());
             assert(self.pipeline.count > 0);
 
-            while (self.pipeline.head_ptr()) |prepare| {
-                assert(self.pipeline.count > 0);
+            // Guard against multiple concurrent invocations of commit_ops()/commit_pipeline():
+            if (self.committing) {
+                log.debug("{}: commit_pipeline: already committing...", .{self.replica});
+                return;
+            }
+
+            self.committing = true;
+            self.commit_pipeline_next();
+        }
+
+        fn commit_pipeline_next(self: *Self) void {
+            assert(self.committing);
+            assert(self.status == .normal);
+            assert(self.leader());
+
+            if (self.pipeline.head_ptr()) |prepare| {
                 assert(self.commit_min == self.commit_max);
-                assert(self.commit_max + self.pipeline.count == self.op);
-                assert(self.commit_max + 1 == prepare.message.header.op);
+                assert(self.commit_min + 1 == prepare.message.header.op);
+                assert(self.commit_min + self.pipeline.count == self.op);
+                assert(self.journal.has(prepare.message.header));
 
                 if (!prepare.ok_quorum_received) {
                     // Eventually handled by on_prepare_timeout().
                     log.debug("{}: commit_pipeline: waiting for quorum", .{self.replica});
+                    self.commit_ops_done();
                     return;
                 }
 
                 const count = prepare.ok_from_all_replicas.count();
                 assert(count >= self.quorum_replication);
                 assert(count <= self.replica_count);
 
-                self.commit_op(prepare.message);
-
-                assert(self.commit_min == self.commit_max);
-                assert(self.commit_max == prepare.message.header.op);
+                self.commit_op_hooks(prepare.message, commit_pipeline_callback);
+            } else {
+                self.commit_ops_done();
+            }
+        }
 
-                self.message_bus.unref(self.pipeline.pop().?.message);
+        fn commit_pipeline_callback(self: *Self) void {
+            assert(self.committing);
+            assert(self.commit_min <= self.commit_max);
+            assert(self.commit_min <= self.op);
 
-                if (self.replica_count == 1) {
-                    if (self.pipeline.head_ptr()) |head| {
-                        // Write the next message in the queue.
-                        // A cluster-of-one writes prepares sequentially to avoid gaps in the WAL.
-                        self.write_prepare(head.message, .append);
-                        // The loop will wrap around and exit when `!ok_quorum_received`.
-                    }
+            if (self.status == .normal and self.leader()) {
```

Comment:
Can you elaborate? This callback is passed to `commit_op_hooks`, and will be called when `compact` finishes (since we don't currently call `checkpoint` from `Replica`).

Reply:
I was wondering if, during a `commit_op_hooks` chain, it's possible for the status of the replica to change, so that it perhaps calls `prefetch()` and perhaps `commit()` etc. but not `compact()`. Or something along those lines. The invariant to preserve is that all hooks in the chain are all-or-nothing.

Reply:
Gotcha. `commit_pipeline_callback` or `commit_journal_callback` is called after compact finishes for each commit. 
So by the time it reaches this branch, `commit_pipeline_callback` has just finished a `prefetch→commit→compact` chain.

Afaik there is no way to only perform part of the chain (except for crashing mid-way through, of course).